Lnd panic: runtime error: invalid memory address or nil pointer dereference

### Background

When I restart the lnd I am get the below error

**panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x58c50e]**
### Your environment

* version of `lnd`
**lnd version 0.6.1-beta**
* which operating system (`uname -a` on *Nix)
**Linux ubuntu-18 4.15.0-50-generic**
* version of `btcd`, `bitcoind`, or other backend
**bitcoind version  v0.18.0.0**

### Expected behaviour
lnd should restart properly or indicate if something has been corrupted.

### Actual behaviour

lnd start crashes with the below logs.
2019-06-11 05:54:27.460 [TRC] CRTR: node=303331363331346462346562353432643137623961376163366636363531666362333139666433646132373931393764346630396138393066633330336563663264 reported failure when sending htlc
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x58c50e]

goroutine 569 [running]:
crypto/ecdsa.Verify(0xc000285b00, 0xc0001d40a0, 0x20, 0x20, 0xc0107969e0, 0xc010796a00, 0xc000134990)
        /usr/local/go/src/crypto/ecdsa/ecdsa.go:237 +0x4e
github.com/btcsuite/btcd/btcec.(*Signature).Verify(...)
        /root/gocode/pkg/mod/github.com/btcsuite/btcd@v0.0.0-20190605094302-a0d1e3e36d50/btcec/signature.go:77
github.com/lightningnetwork/lnd/routing.VerifyChannelUpdateSignature(0xc0000ce630, 0xc000285b00, 0x0, 0x0)
        /root/gocode/src/github.com/lightningnetwork/lnd/routing/ann_validation.go:155 +0x1dd
github.com/lightningnetwork/lnd/routing.ValidateChannelUpdateAnn(0xc000285b00, 0xf42400, 0xc0000ce630, 0xc0000342c0, 0xc000134630)
        /root/gocode/src/github.com/lightningnetwork/lnd/routing/ann_validation.go:136 +0x70
github.com/lightningnetwork/lnd/routing.(*ChannelRouter).applyChannelUpdate(0xc0000ce420, 0xc0000ce630, 0xc000285b00, 0x7d1979a23dfd19b3)
        /root/gocode/src/github.com/lightningnetwork/lnd/routing/router.go:2077 +0x176
github.com/lightningnetwork/lnd/routing.(*ChannelRouter).processSendError(0xc0000ce420, 0x1457520, 0xc01097f840, 0xc0107c1a50, 0xc010a7e1e0, 0xc0001d26c0)
        /root/gocode/src/github.com/lightningnetwork/lnd/routing/router.go:1949 +0x491
github.com/lightningnetwork/lnd/routing.(*paymentLifecycle).handleSendError(0xc000252230, 0x1438420, 0xc010a7e1e0, 0xc0107967e0, 0x2)
        /root/gocode/src/github.com/lightningnetwork/lnd/routing/payment_lifecycle.go:348 +0x25f
github.com/lightningnetwork/lnd/routing.(*paymentLifecycle).resumePayment(0xc000252230, 0x0, 0x0, 0x0, 0x0, 0x1, 0x10a6480, 0x110f101)
        /root/gocode/src/github.com/lightningnetwork/lnd/routing/payment_lifecycle.go:150 +0x56f
github.com/lightningnetwork/lnd/routing.(*ChannelRouter).sendPayment(0xc0000ce420, 0xc0107c1a40, 0xc0004bf130, 0x1457520, 0xc01097f840, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /root/gocode/src/github.com/lightningnetwork/lnd/routing/router.go:1794 +0x1f6
github.com/lightningnetwork/lnd/routing.(*ChannelRouter).Start.func1(0xc0000ce420, 0xc010a5a730, 0xc0000a10a0)
        /root/gocode/src/github.com/lightningnetwork/lnd/routing/router.go:538 +0xfb
created by github.com/lightningnetwork/lnd/routing.(*ChannelRouter).Start
        /root/gocode/src/github.com/lightningnetwork/lnd/routing/router.go:526 +0x33c

--- a/routing/router.go
+++ b/routing/router.go
@@ -1465,9 +1465,11 @@ func generateSphinxPacket(rt *route.Route, paymentHash []byte,
 
 	log.Tracef("Constructed per-hop payloads for payment_hash=%x: %v",
 		paymentHash[:], newLogClosure(func() string {
-			path := sphinxPath[:sphinxPath.TrueRouteLength()]
+			path := make([]sphinx.OnionHop, sphinxPath.TrueRouteLength())
 			for i := range path {
-				path[i].NodePub.Curve = nil
+				hopCopy := sphinxPath[i]
+				hopCopy.NodePub.Curve = nil
+				path[i] = hopCopy
 			}
 			return spew.Sdump(path)
 		}),
@@ -1491,10 +1493,14 @@ func generateSphinxPacket(rt *route.Route, paymentHash []byte,
 
 	log.Tracef("Generated sphinx packet: %v",
 		newLogClosure(func() string {
-			// We unset the internal curve here in order to keep
-			// the logs from getting noisy.
-			sphinxPacket.EphemeralKey.Curve = nil
-			return spew.Sdump(sphinxPacket)
+			// We make a copy of the ephemeral key and unset the
+			// internal curve here in order to keep the logs from
+			// getting noisy.
+			key := *sphinxPacket.EphemeralKey
+			key.Curve = nil
+			packetCopy := *sphinxPacket
+			packetCopy.EphemeralKey = &key
+			return spew.Sdump(packetCopy)
 		}),
 	)
 
@@ -1719,12 +1725,21 @@ func (r *ChannelRouter) sendPayment(
 
 	log.Tracef("Dispatching route for lightning payment: %v",
 		newLogClosure(func() string {
+			// Make a copy of the payment with a nilled Curve
+			// before spewing.
+			var routeHints [][]zpay32.HopHint
 			for _, routeHint := range payment.RouteHints {
+				var hopHints []zpay32.HopHint
 				for _, hopHint := range routeHint {
-					hopHint.NodeID.Curve = nil
+					h := hopHint.Copy()
+					h.NodeID.Curve = nil
+					hopHints = append(hopHints, h)
 				}
+				routeHints = append(routeHints, hopHints)
 			}
-			return spew.Sdump(payment)
+			p := *payment
+			p.RouteHints = routeHints
+			return spew.Sdump(p)
 		}),
 	)
 

--- a/zpay32/hophint.go
+++ b/zpay32/hophint.go
@@ -29,3 +29,15 @@ type HopHint struct {
 	// CLTVExpiryDelta is the time-lock delta of the channel.
 	CLTVExpiryDelta uint16
 }
+
+// Copy returns a deep copy of the hop hint.
+func (h HopHint) Copy() HopHint {
+	nodeID := *h.NodeID
+	return HopHint{
+		NodeID:                    &nodeID,
+		ChannelID:                 h.ChannelID,
+		FeeBaseMSat:               h.FeeBaseMSat,
+		FeeProportionalMillionths: h.FeeProportionalMillionths,
+		CLTVExpiryDelta:           h.CLTVExpiryDelta,
+	}
+}

